cmd/api/logging/logging.go

```diff
@@ -0,0 +1,178 @@
+			"logging secret values unset. To configure logging update the kubearchive-logging Secret")
+		return func(c *gin.Context) {
+			abort.Abort(c, loggingCredsErr, http.StatusBadRequest)
+		}
+	}
+	return func(c *gin.Context) {
+		c.Set(userKey, user)
+		c.Set(passwordKey, password)
+	}
+}
+
+// LogRetrieval retrieves a middleware function that checks the logging config and when set
+// expects to find a log url in the context from where retrieve and parse logs
+// It should be called when user, password and logURL are set in the context.
+func LogRetrieval() gin.HandlerFunc {
+	// FIXME For now the queries to the logging backend server are done insecurely. Needed for the current test env.
+	// #nosec G402
+	client := http.Client{
+		Transport: otelhttp.NewTransport(&http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true}})}
+	return func(c *gin.Context) {
+		user := c.GetString(userKey)
+		password := c.GetString(passwordKey)
+		if user == "" || password == "" {
+			loggingCredsErr := fmt.Errorf(
+				"unexpected error. Logging credentials are unset")
+			abort.Abort(c, loggingCredsErr, http.StatusInternalServerError)
+			return
+		}
+		logUrl := c.GetString("logURL")
+		jsonPath := c.GetString("jsonPath")
+
+		if logUrl == "" {
+			abort.Abort(c, fmt.Errorf("no log URL found"), http.StatusNotFound)
+			return
+		}
+		slog.Info("Retrieving logs", "logURL", logUrl)
+
+		var jsonPathParser jp.Expr
+		var errJsonPath error
+
+		if jsonPath != "" {
+			jsonPathParser, errJsonPath = jp.ParseString(jsonPath)
+			if errJsonPath != nil {
+				abort.Abort(c, fmt.Errorf("invalid jsonPath %s for url %s", jsonPath, logUrl),
+					http.StatusInternalServerError)
+				return
+			}
+		}
+
+		request, errReq := http.NewRequest(http.MethodGet, logUrl, nil)
+		if errReq != nil {
+			abort.Abort(c, errReq, http.StatusInternalServerError)
+			return
+		}
+
+		request.SetBasicAuth(user, password)
+
+		response, errReq := client.Do(request)
+		if errReq != nil {
+			abort.Abort(c, errReq, http.StatusInternalServerError)
+			return
+		}
+
+		defer response.Body.Close()
+
+		reader := bufio.NewReader(response.Body)
+		line, readErr := readLine(reader)
+		var logsReturned bool
+		for readErr == nil {
+			parsedLine, errParseLine := parseLine(line, jsonPathParser)
+			if errParseLine != nil {
+				abort.Abort(c, errParseLine, http.StatusInternalServerError)
+				return
+			}
+			if parsedLine != "" {
+				logsReturned = true
+				c.String(http.StatusOK, parsedLine+"\n")
+			}
+			line, readErr = readLine(reader)
+		}
+
+		if !logsReturned {
+			abort.Abort(c, fmt.Errorf("no parsed logs for the json path: %s", jsonPathParser.String()),
+				http.StatusNotFound)
+		}
+	}
+}
+
+func parseLine(line []byte, jsonPathParser jp.Expr) (string, error) {
+	if jsonPathParser == nil {
+		return string(line), nil
+	}
+
+	var jsonLine any
+	var errJson error
+	var result string
+
+	jsonLine, errJson = oj.Parse(line)
+	if errJson != nil {
+		return result, errJson
```

Comment:
Why return `result` on error and not `nil`?

Reply:
`result` is a string, so it can't be `nil` (not all types can be `nil` in golang, the default/empty value of a string is `""` (empty string), while for a bool is `false` and for an int is `0`. More details [here](https://go.dev/tour/basics/12)

BTW, at that point in the function, `result` is going to be `""` (string zero value). If you think that explicitly returning an empty string makes things clear I can replace it.

Reply:
No, that's fine. I didn't know that a string couldn't be `nil`.